### ISSUE TITLE
Branching Decisions in a Parser (Roll back position by 1 char?)

### ISSUE THREAD

**Issue by allyjweir** [sprache/Sprache#64]
Hi all,

**TL;DR:** Can I put the position of the parser back by one character so that I can make a decision of which parser to use?

**Full story:** I am working to parse .dfm files which define the layout of old Delphi application windows/forms. I am successfully parsing the objects that define things like buttons on screen or input fields. These each have properties which can either be strings or string arrays.

For example, 
String example -> ```Caption = "Hello"``` 
Array example -> ```Items = ["dave", "gill", "bill"]```

I want to create a generic parser that can choose if a given property assignment is a string or array one based on the first character (if an array, this would be a `[`,`{`, or `(`) then parse an array or string assignment based on this (in both cases returning a List<string>).

**Is there a way to parse the first character of the assignment, have this as `foo` then roll the position of the parser back by 1 character then in the next line of the LINQ statement, feed foo into a function that returns the correct parser based on the contents of `foo`?**

I imagine the code looking like:

```
from foo in Parse.Char.Once()
// Roll back position by one here
from bar in StringOrArrayParser(foo)
select bar  // This returning a List<string> (return type of string parser and array parser)
```

Also I'm pretty new to C# and LINQ (loving Sprache so far, had only lightly played with Parsec previously) so if I am making any huge mistakes I would appreciate direction on those as well.

**Comment by mpartipilo** [comment]
You can't move the current position of the parser, but your problem can be solved without the need to do that.

Since you're dealing with 2 different types, you need to make sure that both parsers encapsulate the return value in a single type.

You can create a base object to encapsulate an array or a string with separate constructors/derived classes and return that base object from the parsers.

So lets look at some notepad (not compiled) code:


```
class SomeValue<T>
{
    T _value = null;
    IEnumerable<T> _values = null;
    SomeValue<T>(T value) { _value = value;}
    SomeValue<IEnumerable<T>>(IEnumerable<T> value) { _values = values;}
}

Parser<string> QuotedText =
            (from lquot in Parse.Char('"')
             from content in Parse.CharExcept('"').Many().Text()
             from rquot in Parse.Char('"')
             select content).Token();

var StringParser = 
   from t in Parse.QuotedText.Text()
   select new SomeValue<string>(t);

var ArrayParser = 
    from a in StringParser.Many().DelimitedBy('[', ']')
    select new SomeValue<string>(a);

var StringOrArrayParser = 
    from x in StringParser.Or(ArrayParser)
    select x;
```

Some extra work is needed for escaping strings with quotes and properly keeping character ']' from ending your parsing earlier than expected.

**Comment by allyjweir** [comment]
Hi @mpartipilo, thanks for your help. I had a lot of that structure already but having it confirmed as the approach to take and more abstractly discussed/viewed without the context I am working in made the changes I had to make much clearer.

It came down to a misunderstanding of the Or/XOr methods and other minor things. [This other issue in the project helped with that to an extent](https://github.com/sprache/Sprache/issues/46)

I appreciate the help, thanks again!
